### PR TITLE
Add options to optimize for size/speed

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
Publicly exposing the undocumented switches, adding test legs.

Cc @dotnet/ilc-contrib @dotnet/jit-contrib

### CODE REVIEW

**Review by jkotas — APPROVED** [review]

**Comment by jkotas** [comment]
This will need a matching PR in the docs repo.

**Comment by MichalStrehovsky** [comment]
> This will need a matching PR in the docs repo.

Typing it up as we speak :)

**Comment by MichalStrehovsky** [comment]
Docs update: dotnet/docs#35092

File: src/coreclr/nativeaot/docs/optimizing.md [hunk 1]
```diff
@@ -37,8 +37,8 @@ Since `PublishTrimmed` is implied to be true with Native AOT, some framework fea
 * `<IlcGenerateStackTraceData>false</IlcGenerateStackTraceData>`: this disables generation of stack trace metadata that provides textual names in stack traces. This is for example the text string one gets by calling `Exception.ToString()` on a caught exception. With this option disabled, stack traces will still be generated, but will be based on reflection metadata alone (they might be less complete).
 
 ## Options related to code generation
-* `<IlcOptimizationPreference>Speed</IlcOptimizationPreference>`: when generating optimized code, favor code execution speed.
-* `<IlcOptimizationPreference>Size</IlcOptimizationPreference>`: when generating optimized code, favor smaller code size.
+* `<OptimizationPreference>Speed</OptimizationPreference>`: when generating optimized code, favor code execution speed.
+* `<OptimizationPreference>Size</OptimizationPreference>`: when generating optimized code, favor smaller code size.
```
[the remark this answers is not in the document]
Comment (MichalStrehovsky) [reply src/coreclr/nativeaot/docs/optimizing.md:41]: It was considered. I've forwarded you the email with discussion.

**Comment by MichalStrehovsky** [comment]
This found a pre-existing issue in casting logic. I've activeissue'd it and not going to block this on that.

**Review by vitek-karas — APPROVED** [review]
I'm only slightly nervous about adding yet more extra-platforms legs
Have you considered replacing one of the existing runs for NativeAOT with one of these (or just run it with opt on).

**Comment by MichalStrehovsky** [comment]
> Have you considered replacing one of the existing runs for NativeAOT with one of these (or just run it with opt on).

I have, but it would be a test hole. It is already a test hole that we don't test any of this on e.g. Mac. We need to keep testing the mainline scenarios because the mainline matrix is _already_ a compromise with holes. What we should consider is:

* Separate test leg for NativeAOT testing. The fact that we run these _consistently failing_ Mono test legs whenever we want to test NativeAOT is the main source of waste. I have not seen the Mono legs green in a year. I am also getting tired of digging through swathes of Mono leg failures just to find the one relevant Native AOT failure. The infra team doesn't optimize for this experience (and I'm not surprised they don't), but being in this "special class" makes build analysis worthless.
* A weekly test leg for NativeAOT. We have a number of test holes already and the test holes are just growing. We need a leg that can tell us we're good for a release.

**Comment by akoeplinger** [comment]
I'd recommend the first option, we have something similar with e.g. eng/pipelines/extra-platforms/runtime-extra-platforms-ioslike.yml which is available both as a standalone pipeline `/azp run runtime-ioslike` as well as included in the runtime-extra-platform.yml

**Comment by MichalPetryka** [comment]
Should the default behaviour that happens when nothing is specified be also documented?

**Comment by MichalStrehovsky** [comment]
> Should the default behaviour that happens when nothing is specified be also documented?

It's documented in https://github.com/dotnet/docs/pull/35092 (that's where docs are). Is there something you'd like to see changed?

**Pull request merged**